src/carousel/tp-carousel.ts

```diff
@@ -0,0 +1,413 @@
+/**
+ * Internal dependencies.
+ */
+import { TPCarouselArrowElement } from './tp-carousel-arrow';
+import { TPCarouselSlideElement } from './tp-carousel-slide';
+import { TPCarouselSlidesElement } from './tp-carousel-slides';
+
+/**
+ * TP Carousel.
+ */
+export class TPCarouselElement extends HTMLElement {
+	protected _observer: IntersectionObserver;
+	protected slides: NodeListOf<TPCarouselSlideElement> | null;
+	protected slideTrack: TPCarouselSlidesElement | null;
+	protected isProgramaticScroll: boolean = false;
+
+	/**
+	 * Constructor.
+	 */
+	constructor() {
+		// Initialize parent.
+		super();
+		this.slides = this.querySelectorAll( 'tp-carousel-slide' );
+		this.slideTrack = this.querySelector( 'tp-carousel-slides' );
+		this._observer = new IntersectionObserver( this.attributeChangeOnScroll?.bind( this ), { root: this.slideTrack, threshold: 0.999 } );
```

Comment:
@harshdeep-gill Do we need to be that strict to use 0.999 for 100% visibility of the element, or 0.99 should be enough?
Being too strict might not trigger if the root container clips a pixel or two due to rounding or scroll position.